File: TCP-UDP/TCP_Client.c
/************************************
*     Ricatti Luigi Francesco       *
*     TCP_Client.c                  *
*           		                    *
************************************/

/* Client Stream versione 1: una socket per ogni richiesta (dentro al ciclo while)
 * il client fa una nuova connect ad ogni ciclo */

/******** #INCLUDE ********/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>

/******** #DEFINE *******/
#define DIM_BUFF 256


int main(int argc, char *argv[])
{
      int sd, nread, nwrite, port, num;
      char buff[DIM_BUFF];
      struct hostent *host;
      struct sockaddr_in servaddr;

    /* CONTROLLO ARGOMENTI ---------------------------------- */
      if(argc!=3)
      {
	printf("Error:%s serverAddress serverPort\n", argv[0]);
	exit(1);
      }
      /* PREPARAZIONE INDIRIZZO SERVER ----------------------------- */
      memset((char *)&servaddr, 0, sizeof(struct sockaddr_in));
      servaddr.sin_family = AF_INET;
      host = gethostbyname(argv[1]);

      // Controllo host
      if (host == NULL)
      {
	printf("%s not found in /etc/hosts\n", argv[1]);
	exit(1); 
      }
      // Controllo che la porta sia un intero
      num=0;
	    while( argv[2][num]!= '\0' )
	    {
		  if( (argv[2][num] < '0') || (argv[2][num] > '9') )
		  {
			  printf("Argomento non intero\n");
			  exit(2);
		  }
		  num++;
	    }

      // Controllo che la porta sia nel range
      port = atoi(argv[2]);
      if (port < 1024 || port > 65535)
      {
	printf("Error: porta non valida!\n");
	printf("1024 <= port <= 65535\n");
	exit(2);
      }
       else
  {
    servaddr.sin_addr.s_addr=((struct in_addr*) (host->h_addr))->s_addr;
    servaddr.sin_port = htons(atoi(argv[2]));
  }
  
  printf("Client avviato\n");
  
  
  /* CORPO DEL CLIENT: */
  /* ciclo di accettazione di richieste da utente ------- */
  printf("\n************************************\n");  
  printf("Richieste dell'utente, EOF per terminare: ");
  
  // setto la condizione per il ciclo
  int condition =0;
  while(condition)
  {
    /* CREAZIONE E CONNESSIONE SOCKET (BIND IMPLICITA) ----------------- */
    /* in questo schema è necessario ripetere creazione, settaggio opzioni e connect */
    /* ad ogni ciclo, perchè il client fa una nuova connect ad ogni ciclo */

    sd=socket(AF_INET, SOCK_STREAM, 0);
    if (sd <0)
    {
      perror("apertura socket "); 
      exit(3);
    }
    printf("Creata la socket sd=%d\n", sd);

    if (connect(sd,(struct sockaddr *) &servaddr, sizeof(struct sockaddr))<0)
    {
      perror("Errore in connect");
      exit(4);
    }
    printf("Connect ok\n");

    //qui logica applicativa
    printf("\n************************************\n");

    // Write su socket
    printf("Invio richiesta\n");
    while((nwrite=write(sd, buff, DIM_BUFF))>0)
    {
      if(nwrite < 0
      {
	printf("Errore nella scrittura su socket!");
	close(sd);
	exit(7);
      }
      else
      {
	write(1,buff,nwrite);
	printf("ho scritto su socket %d byte", nwrite);
      }
    }
    printf("Richiesta inviata\n");

    // Chiusura socket in spedizione -> invio dell'EOF
    shutdown(sd,1);

    // Read dalla socket
    printf("Ricevo la risposta:\n");
    while((nread=read(sd,buff,DIM_BUFF))>0)
    {
      if(nread < 0)
      {
	printf("Errore nella lettura su socket!");
	lose(sd);
	exit(6);
      }
      else
      {
	write(1,buff,nread);
	printf("ho letto da socket %d byte", nwrite);
      }
    }
    printf("Risposta ricevuta\n");
    
    // Chiusura socket in ricezione
    shutdown(sd, 0);
    close(sd); // Close fuori dal ciclo per client con una connessione
        
    printf("Nuova richiesta, EOF per terminare: ");
    
  }//while
  
  printf("\nClient: termino...\n");
  exit(0);
  
}//main
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
      

File: TCP-UDP/TCP_Client.java
/************************************
*     Ricatti Luigi Francesco       *
*     TCP_Client.java	            *
*           		            *
************************************/


/* CLIENTE CON CONNESSIONE*/

import java.net.*;
import java.io.*;

public class TCP_Client 
{

  /**
  * @param args - Usage: java TCP_Client <address> <port>
  */
  
 	public static void main(String[] args) throws IOException 
  	{
	     /*
	     * Come argomenti si devono passare un nome di host valido e una porta
	     *  
	     */
	     InetAddress addr = null;
	     int port = -1;

	      // Controllo argomenti 
	     try
	     {
	     	if(args.length == 2)
	     	{
	     		addr = InetAddress.getByName(args[0]);
	     		port = Integer.parseInt(args[1]);
	     	} 
	     	else
	     	{
	     		System.out.println("Usage: java TCP_Client serverAddr serverPort");
	     		System.exit(1);
	     	}
	     } 
	     catch(NumberFormatException e)
	     {
	     	System.out.println("Numero di porta errato: " + args[1]);
	     	e.printStackTrace();
	     	System.out.println("Usage: java TCP_Client serverAddr serverPort");
	     	System.exit(1);
	     }
	     catch(Exception e)
	     {
	     	System.out.println("Problemi, i seguenti: ");
	     	e.printStackTrace();
	     	System.out.println("Usage: java TCP_Client serverAddr serverPort");
	     	System.exit(1);
	     }

	     /** ----- Inizializzazione strutture dati Cliente ----- **/
	     Socket socket = null;
	     DataInputStream inSock = null;
	     DataOutputStream outSock = null;
	     String richiesta = null;

	      // Creazione socket 
	      // Creazione stream di input/output su socket
	     try
	     {
	     	socket = new Socket(addr, port);
		    // Setto il timeout per non bloccare indefinitivamente il client
	     	socket.setSoTimeout(30000);
	     	System.out.println("Creata la socket: " + socket);
	     	inSock = new DataInputStream(socket.getInputStream());
	     	outSock = new DataOutputStream(socket.getOutputStream());
	     }
	     catch(IOException e)
	     {
	     	System.out.println("Problemi nella creazione degli stream su socket: ");
	     	e.printStackTrace();
	     	System.exit(1);
	     }
	     catch(Exception e)
	     {
	     	System.out.println("Problemi nella creazione della socket: ");
	     	e.printStackTrace();
	     	System.exit(2);
	     }

	      // Eventuale stream per l'interazione con l'utente da tastiera

	     BufferedReader stdIn = new BufferedReader(new InputStreamReader(System.in));
	     System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");
	     try
	     {

	     	while((richiesta = stdIn.readLine()) != null)
	     	{
		  		/*
			  	if(richiesta.equalsIgnoreCase("U") || richiesta.equalsIgnoreCase("D"))
		      	{
		      		outSock.writeUTF(richiesta);
		      		System.out.println("Inviata richiesta: " + richiesta);
		      	}
		      	outSock.flush();

		      	if(richiesta.equalsIgnoreCase("U"))
		      	{*/
		      		try
		      		{
				      // Elaborazione richiesta e invio
		      			outSock.writeUTF(richiesta);
		      			System.out.println("Inviata richiesta: " + richiesta);
				      // chiudo output della socket 
		      			socket.shutdownOutput();
		      			System.out.println("Chiusura output della socket: " + socket);
		      		}
		      		catch(Exception e)
		      		{
		      			System.out.println("Problemi nell'invio della richiesta " + richiesta );
		      			e.printStackTrace();
		      			System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            
		      			continue;
				      // il client continua l'esecuzione riprendendo dall'inizio del ciclo
		      		}

			    	//--> qui logica 


			    	// Ricezione della risposta

		      		String risposta;
		      		try
		      		{
		      			risposta = inSock.readUTF();
		      			System.out.println("Risposta: " + risposta);
				      // chiudo input della socket
		      			socket.shutdownInput();
		      			System.out.println("Chiusura input della socket: " + socket);
		      		}
		      		catch(SocketTimeoutException ste)
		      		{
		      			System.out.println("Timeout scattato: ");
		      			ste.printStackTrace();
		      			socket.close();
		      			System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            
		      			continue;
				      	// il client continua l'esecuzione riprendendo dall'inizio del ciclo         
		      		}

		      		catch (EOFException e) 
		      		{
		      			System.out.println("Raggiunta la fine delle ricezioni, chiudo...");
		      			socket.close();
		      			System.out.println("TCP_Client: termino...");
		      			System.exit(0);
		      		}
		      		catch(Exception e)
		      		{
		      			System.out.println("Problemi nella ricezione della risposta, i seguenti: ");
		      			e.printStackTrace();
		      			socket.close();
		      			System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            
		      			continue;
				      // il client continua l'esecuzione riprendendo dall'inizio del ciclo
		      		}
			/*	}
				else if(richiesta.equalsIgnoreCase("U"))
		      	{
					.... uguale con try catch come sopra ... e se ci sono altri if continua
				} 
				else System.out.println("Errore nella scelta..."); */

				// Tutto ok, pronto per nuova richiesta
		      	System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");	

			}// while

			System.out.println("TCP_Client: termino...");
		}

	    // qui catturo le eccezioni non catturate all'interno del while
	    // quali per esempio la caduta della connessione con il server
	    // in seguito alle quali il client termina l'esecuzione
		catch(Exception e)
		{
			System.err.println("Errore irreversibile, il seguente: ");
			e.printStackTrace();
			System.err.println("Chiudo!");
			System.exit(3); 
		}
	} // main
}


/* CLIENTE CON MULTICONNESSIONE*/


/************************************
*     Ricatti Luigi Francesco       *
*     TCP_Client.java	            *
*           		            *
************************************/

import java.net.*;
import java.io.*;

public class TCP_Client 
{
  public static void main(String[] args) throws IOException 
  {
    // Inizializzazione argomenti
    InetAddress addr = null;
    int port = -1;

    // Controllo argomenti 
    try
    {
      if(args.length == 2)
      {
	addr = InetAddress.getByName(args[0]);
	port = Integer.parseInt(args[1]);
      } 
      else
      {
	System.out.println("Usage: java TCP_Client serverAddr serverPort");
	System.exit(1);
      }
    } 
    catch(NumberFormatException e)
    {
      System.out.println("Numero di porta errato: " + args[1]);
      e.printStackTrace();
      System.out.println("Usage: java TCP_Client serverAddr serverPort");
      System.exit(1);
    }
    catch(Exception e)
    {
      System.out.println("Problemi, i seguenti: ");
      e.printStackTrace();
      System.out.println("Usage: java TCP_Client serverAddr serverPort");
      System.exit(1);
    }

    // Inizializzazione strutture dati Cliente
    Socket socket = null;
    DataInputStream inSock = null;
    DataOutputStream outSock = null;
    String richiesta = null;

    // Eventuale stream per l'interazione con l'utente da tastiera
    BufferedReader stdIn = new BufferedReader(new InputStreamReader(System.in));
    System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");

    try
    {
      while((richiesta = stdIn.readLine()) != null)
      {  
	// Creazione socket
	// Creazione stream di input/output su socket
	try
	{
	  socket = new Socket(addr, port);
	  // Setto il timeout per non bloccare indefinitivamente il client
	  socket.setSoTimeout(30000);
	  System.out.println("Creata la socket: " + socket);
	  inSock = new DataInputStream(socket.getInputStream());
	  outSock = new DataOutputStream(socket.getOutputStream());
	}
	catch(IOException e)
	{
	  System.out.println("Problemi nella creazione degli stream su socket: ");
	  e.printStackTrace();
	  System.out.println("Nuova richiesta: ");
	  System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            
	  continue;
	  // il client continua l'esecuzione riprendendo dall'inizio del ciclo
	}
	catch(Exception e)
	{
	  System.out.println("Problemi nella creazione della socket: ");
	  e.printStackTrace();
	  System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");
	  continue;
	  // il client continua l'esecuzione riprendendo dall'inizio del ciclo
	}
	

	// Elaborazione e trasmissione della richiesta
	try
	{
	  // Elaborazione richiesta e invio
	  outSock.writeUTF(richiesta);
	  System.out.println("Inviata richiesta: " + richiesta);
	  // chiudo output della socket 
	  socket.shutdownOutput();
	  System.out.println("Chiusura output della socket: " + socket);
	}
	catch(Exception e)
	{
	  System.out.println("Problemi nell'invio della richiesta " + richiesta );
	  e.printStackTrace();
	  System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            
	  continue;
	  // il client continua l'esecuzione riprendendo dall'inizio del ciclo
	}

	// Ricezione della risposta
	String risposta;
	try
	{
	  risposta = inSock.readUTF();
	  System.out.println("Risposta: " + risposta);
	  // chiudo input della socket
	  socket.shutdownInput();
	  System.out.println("Chiusura input della socket: " + socket);
	}
	catch(SocketTimeoutException ste)
	{
	  System.out.println("Timeout scattato: ");
	  ste.printStackTrace();
	  socket.close();
	  System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            
	  continue;
	  // il client continua l'esecuzione riprendendo dall'inizio del ciclo         
	}
	catch (EOFException e) 
	{
	  System.out.println("Raggiunta la fine delle ricezioni, chiudo...");
	  socket.close();
	  System.out.println("GetFileClient: termino...");
	  System.exit(0);
	}
	catch(Exception e)
	{
	  System.out.println("Problemi nella ricezione della risposta, i seguenti: ");
	  e.printStackTrace();
	  socket.close();
	  System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            
	  continue;
	  // il client continua l'esecuzione riprendendo dall'inizio del ciclo
	}

	// Tutto ok, pronto per nuova richiesta
	System.out.print("\n^D(Unix)/^Z(Win)+invio per uscire, oppure immetti input: ");		            

      }// while
      
      System.out.println("TCP_Client: termino...");
    }
    // qui catturo le eccezioni non catturate all'interno del while
    // quali per esempio la caduta della connessione con il server
    // in seguito alle quali il client termina l'esecuzione
    
    catch(Exception e)
    {
      System.err.println("Errore irreversibile, il seguente: ");
      e.printStackTrace();
      System.err.println("Chiudo!");
      System.exit(3); 
    }

  } // main
}

*/
